tools/module_diag_hailcast.F90

```diff
@@ -0,0 +1,1870 @@
+      !!!  COLLECTION EFFICIENCY FOR WATER AND ICE
+      !EW=1.0
+      !!!!   IF TS WARMER THAN -5C THEN ACCRETE ALL THE ICE (EI=1.0)
+      !!!!   OTHERWISE EI=0.21
+      !IF(TS.GE.268.15)THEN
+      !  EI=1.00
+      !ELSE
+      !  EI=0.21
+      !ENDIF
+
+      !!!  COLLECTION EFFICIENCY FOR WATER AND ICE
+      EW=1.0
+      !!!  Linear function for ice accretion efficiency
+      IF (TC .GE. 273.155) THEN
+         EI=1.00
+      ELSE IF (TC.GE.233.155) THEN
+         EI= 1.0 - ( (273.155 - TS) / 40. )
+      ELSE  !cooler than -40C
+         EI = 0.0
+      ENDIF
+
+      !!!  CALCULATE THE VENTILATION COEFFICIENT - NEEDED FOR GROWTH FROM VAPOR
+      !The coefficients in the ventilation coefficient equations have been
+      !experimentally derived, and are expecting cal-C-g units.  Do some conversions.
+      DENSAC = DENSA * (1.E3) * (1.E-6)
+      !dynamic viscosity
+      ANU=1.717E-4*(393.0/(TC+120.0))*(TC/273.155)**1.5
+      !!!  CALCULATE THE REYNOLDS NUMBER - unitless
+      RE=D*VT*DENSAC/ANU
+      E=(0.60)**(0.333333333)*(RE**0.50) !ventilation coefficient vapor (fv)
+      !!!   SELECT APPROPRIATE VALUES OF AE ACCORDING TO RE
+      IF(RE.LT.6000.0)THEN
+         AE=0.78+0.308*E
+      ELSEIF(RE.GE.6000.0.AND.RE.LT.20000.0)THEN
+         AE=0.76*E
+      ELSEIF(RE.GE.20000.0) THEN
+         AE=(0.57+9.0E-6*RE)*E
+      ENDIF
+
+
+      !!!  CALC HAILSTONE'S MASS (GM), MASS OF WATER (GMW) AND THE
+      !!!  MASS OF ICE IN THE STONE (GMI)
+      GM=PI/6.*(D**3.)*DENSE
+      GMW=FW*GM
+      GMI=GM-GMW
+
+      !!!  STORE THE MASS
+      GM1=GM
+
+      !!! NEW MASS GROWTH CALCULATIONS WITH VARIABLE RIME
+      !!! LAYER DENSITY BASED ON ZIEGLER ET AL. (1983)
+
+      !!! CALCULATE INCREASE IN MASS DUE INTERCEPTED CLD WATER, USE
+      !!! ORIGINAL DIAMETER
+      GMW2=GMW+SEKDEL*(PI/4.*D**2.*VT*XW*EW)
+      DGMW=GMW2-GMW
+      GMW=GMW2
+
+      !!!  CALCULATE THE INCREASE IN MASS DUE INTERCEPTED CLOUD ICE
+      GMI2=GMI+SEKDEL*(PI/4.*D**2.*VT*XI*EI)
+      DGMI=GMI2-GMI
+      GMI=GMI2
+
+      !!! CALCULATE INCREASE IN MASS DUE TO SUBLIMATION/CONDENSATION OF
+      !!! WATER VAPOR
+      DGMV = SEKDEL*2*PI*D*AE*DI*DELRW
+      IF (DGMV .LT. 0) DGMV=0
+
+      !!!  CALCULATE THE TOTAL MASS CHANGE
+      DGM=DGMW+DGMI+DGMV
+      !Dummy arguments in the event of no water, vapor, or ice growth
+      DENSELW = 900.
+      DENSELI = 700.
+      !!! CALCULATE DENSITY OF NEW LAYER, DEPENDS ON FW AND ITYPE
+      IF (ITYPE.EQ.1) THEN !DRY GROWTH
+          !If hailstone encountered supercooled water, calculate new layer density
+          ! using Macklin form
+          IF ((DGMW.GT.0).OR.(DGMV.GT.0)) THEN
+             !MEAN CLOUD DROPLET RADIUS, ASSUME CLOUD DROPLET CONC OF 3E8 M-3 (300 CM-3)
+             DC = (0.74*XW / (3.14159*1000.*3.E8))**0.33333333 * 1.E6 !MICRONS
+             !!! FIND THE STOKES NUMBER  (rasmussen heymsfield 1985)
+             NS = 2*VT*100.*(DC*1.E-4)**2. / (9*ANU*D*50)  !need hail radius in cm
```

Comment:
I suggest replacing these exponent expressions with multiplications if possible, e.g. ((DC*1.E-4)**2. ->  (DC*1.E-4)*(DC*1.E-4), also W**2->W*W, W**3->W*W*W, w**4->W*W*W*W for computational efficiency.